src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestRunConfigurationProducer.kt

```diff
@@ -70,26 +70,42 @@ class CargoTestRunConfigurationProducer : RunConfigurationProducer<CargoCommandC
 
     private fun findTestFunction(location: Location<*>): TestConfig? {
         val fn = location.psiElement.parentOfType<RsFunction>(strict = false) ?: return null
+        val testName = fn.testName() ?: return null
         val name = fn.crateRelativePath.configPath() ?: return null
         val target = fn.containingCargoTarget ?: return null
-        return if (fn.isTest) TestConfig(fn, "Test $name", name, target) else null
+        return TestConfig(fn, testName, name, target)
     }
 
     private fun findTestMod(location: Location<*>): TestConfig? {
         val mod = location.psiElement.parentOfType<RsMod>(strict = false) ?: return null
-        val testName = if (mod.modName == "test" || mod.modName == "tests")
-            "Test ${mod.`super`?.modName}::${mod.modName}"
-        else
-            "Test ${mod.modName}"
+        val testName = mod.testName() ?: return null
 
         val testPath = mod.crateRelativePath.configPath() ?: ""
         val target = mod.containingCargoTarget ?: return null
-        if (!mod.functionList.any { it.isTest }) return null
 
         return TestConfig(mod, testName, testPath, target)
     }
 }
 
+fun PsiElement.testName(): String? {
```

Comment:
I would rather try to avoid such all encompassing extensions methods: now every PSI element anywhere in the project has `testName` method. 

What do you think about the design similar to that of `ExecutableRunLineMarkerContributor`? https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/cargo/runconfig/command/CargoExecutableRunLineMarkerContributor.kt#L21

There, line marker calls into `CargoExecutableRunConfigurationProducer` which acts like a single source of truth for test discovery. Here, we won't be able to use a simple `Boolean` is test function, because we need a more rich return type with two variants "Nope, that's not a test!" and "Yep, it's a test, and it has a name "foo_test" ."